### ISSUE TITLE
Longest prefix lookup?

### ISSUE THREAD

**Issue by vdukhovni** [Deewiant/list-tries#10]
Do (or could) any of the data structures in this package support longest-prefix lookup?  That is, if my key is [a,b,c,d,e] and the map has an entry with key [a,b,c] (but not [a,b,c,d] or [a,b,c,d,e]) I'd like a lookup function that returns the value for [a,b,c].  I did not see that listed in the docs, but it seems like a very natural query for maps with lists as keys...

**Comment by Deewiant** [comment]
I believe `findPredecessor` is an equivalent operation? E.g.:

```hs
import qualified Data.ListTrie.Patricia.Set.Enum as LT
s :: LT.TrieSet Char
s = LT.fromList ["abc","ab","abb","abd"]
abc = LT.findPredecessor "abcde" s -- Just "abc"
```

**Comment by vdukhovni** [comment]
But a *predecessor* is not necessarily a *prefix*.  For example:
```
import qualified Data.ListTrie.Patricia.Set.Enum as LT

s :: LT.TrieSet Char
s = LT.fromList ["abc","ab","abb","abd"]

abc :: Maybe String
abc = LT.findPredecessor "ac" s -- Just "abd"
```
I am looking for the longest prefix that has a entry in the map (not an intermediate node).

**Comment by Deewiant** [comment]
Right, my bad. I was thinking that you could filter those cases out by checking whether the result is a prefix of your key, but that doesn't work for e.g. `findPredecessor "ac" (fromList ["a", "ab"])` which will give no indication of the `"a"`.

You should be able to construct the function you need using `splitPrefix` and `children` since they let you traverse the trie structure manually.

Otherwise, PRs are welcome. Unfortunately I don't have much time to spend on list-tries so I can't give any realistic ETA on when I might get around to this myself.

**Comment by vdukhovni** [comment]
I am looking for a somewhat efficient function that will return the node with the longest common prefix in a single pass, rather than multiple searches through the tree.  So this would require new code. Not sure I'll step up to do that, but I'll consider it.

**Comment by Deewiant** [comment]
Fair enough. Note though that it should be possible to write a fairly straightforward single-pass implementation with only `splitPrefix` and `children`.

**Comment by vdukhovni** [comment]
Thanks.  I think I see what you mean.  I'll give it a go...
